src/ml/trainer/JasmineGraphTrainingSchedular.cpp

```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
+ *
+ * @param partitionDetails a vector of vectors containing partition ID, node count, edge count, feature count per partition
+ * @param availableMemory total memory of host
+ */
+vector<pair<int, int>>
+JasmineGraphTrainingSchedular::estimateMemoryDistOpt(vector<vector<int>> partitionDetails, long availableMemory) {
+
+
+    vector<pair<int, int>> partitionMemoryList; //Vector of estimated sizes of partitions
+
+    //Iterate through partitions and estimate sizes
+    for (vector<vector<int>>::iterator i = partitionDetails.begin(); i != partitionDetails.end(); i++) {
+        int partitionID = (*i)[0];
+        int nodeCount = (*i)[1];
+        int edgeCount = (*i)[2];
+        int featureCount = (*i)[3];
+
+        int graphSize = 8*nodeCount*edgeCount*featureCount / (1024 * 1024); //Size in hard disk
+
+        int sizeInMem = 36 * graphSize + 2; //Size when loaded into memory
+
+        if (sizeInMem > availableMemory) {
+            sizeInMem = availableMemory;
+        }
+
+        partitionMemoryList.push_back(make_pair(partitionID, sizeInMem));
+    }
+
+    return partitionMemoryList;
+}
+
+/** Create schedule for host by best fitting in sorted order (descending)
+ *
+ * @param partitionMemoryList Vector of pairs containing partition ID and estimated size in memory
+ * @param partitionWorkerMap Map of partition id and the worker assigned to partition
+ * @param capacity total memory of host
+ */
+std::map<int, map<int, int>>
+JasmineGraphTrainingSchedular::bestFitPartition(vector<pair<int, int>> partitionMemoryList, map<int, int> partitionWorkerMap, int capacity) {
+
+    std::map<int, map<int, int>> schedule; //Host schedule per worker and what partitions to load in which order
+
+    //Initialize the state of host workers as free
+    std::map<int, bool> workerState;
+    for (auto i: partitionWorkerMap) workerState[i.second] = true;
+
+    //Sort partition memory list in descending order
+    sort(partitionMemoryList.begin(), partitionMemoryList.end(),
+         [](pair<int, int> &a, pair<int, int> &b) { return (a.second > b.second); });
+
+
+    int order = 0; //Order of partitions being loaded into memory
+    int remainingMem = capacity; //Total remaining mem after each pass (iteration) in simulation
+    vector<pair<int, int>> simulatedMem; //Partitions in mem and their remaining work (in memory) in simulation
+
+    while (schedule.size() < partitionMemoryList.size()) {
+        int proceedToNextIter = 0;
+        //Iterate through sorted partition list and assign to memory (best fit)
+        for (vector<pair<int, int>>::iterator i = partitionMemoryList.begin(); i != partitionMemoryList.end(); i++) {
+            int partition = i->first;
+            int mem = i->second;
+            int worker = partitionWorkerMap[partition];
+
+            //Check that partition is not already scheduled
+            if (schedule.find(partition) == schedule.end()) continue;
```

Comment:
Shouldn't this be `if (schedule.find(partition) != schedule.end()) continue;` instead? According to the [documentation](http://www.cplusplus.com/reference/map/map/find/) `find()` returns an iterator if it finds the key in the map if not returns the `end()`. Therefore, if the partition is already scheduled `find()` will not return `end()`. Hence we should check that condition and continue to the top of the loop.

Reply:
Done the changes as suggested and also included some additional condition checks required for this purpose.

Reply:
Done